### PR TITLE
[glsl-out] Reworked the backend

### PR DESCRIPTION
**Pull request by JCapucho**
The backend now works in a similar way to the msl backend
Should require less loops, allocations and backtracking
Overall just better to work with it

Added a build script which reads a file with the glsl keywords and
generates a slice to be used in the Namer

Added a way to build a call graph

### CODE REVIEW

**Comment by JCapucho** [comment]
Completely forgot to mention but right now this design requires ordered structs and ignores builtin members.
But all tests pass

File: src/back/glsl.rs [hunk 1]
```diff
@@ -282,1180 +314,1122 @@ impl FeaturesManager {
...
-                            members[index as usize]
-                                .name
-                                .as_ref()
-                                .filter(|s| is_valid_ident(s))
-                                .unwrap_or(&format!("_{}", index))
-                        ))
+    fn write_expr(
+        &mut self,
+        expr: Handle<Expression>,
+        ctx: &mut FunctionCtx<'_, '_>,
```
Comment (kvark) [inline src/back/glsl.rs:1043]: does it need `&mut` here?

**Review by kvark — COMMENTED** [review]
Grand work, and a good reduction in line count, together with better code all around!
I have a few concerns, overall very happy to get this landed 🥇

File: src/back/glsl.rs [hunk 2]
```diff
@@ -282,1180 +314,1122 @@ impl FeaturesManager {
...
+            return Err(Error::Custom(format!(
+                "Version not supported {}",
+                options.version
+            )));
+        }
 
-    if let Some(depth_test) = entry_point.early_depth_test {
-        manager.request(Features::IMAGE_LOAD_STORE);
-        writeln!(&mut buf, "layout(early_fragment_tests) in;\n")?;
+        let (ep_idx, ep) = module
```
[the remark this answers is not in the document]
Comment (JCapucho) [reply src/back/glsl.rs:383]: I don't know how this can be possible since the entry point index is used in the namer to get locals names

**Comment by JCapucho** [comment]
@kvark thanks for the great review, i managed to further clean up and reduce more the line count and the loops while working on the comments shouldn't be anything to big

**Review by kvark — APPROVED** [review]

**Comment by kvark** [comment]
Thank you, amazing work here!

**Pull request merged**